Add support for WebDriver cookie operations

aksiksi opened the pull request:
## Overview

This PR adds support for the cookie operations defined in [16. Cookies](https://www.w3.org/TR/webdriver1/#cookies) in the WebDriver spec. In particular, I've implemented the following operations:

* Get All Cookies
* Get Named Cookie
* Delete Cookie
* Delete All Cookies

The last operation, "Add Cookie", requires support from the `webdriver` crate.  The primary issue is that the `AddCookieParameters` struct does not implement `Serialize`, which means that we can't build the POST request on our side.  I'll be trying to get in touch with the geckodriver folks to see if this can be changed on their end.

## Changes

* Added a new module, `cookies.rs`.  This module contains the cookie-related `Client` methods as well as a few functions for converting between raw JSON (from WebDriver) and `cookie::Cookie` for the APIs.
* Added a simple integration testcase that triggers the `get_all_cookies` and `delete_all_cookies` operations.
* Removed the `binary` flag/arg when creating the WebDriver test client.  Both `chromedriver` and `geckodriver` are able to detect the path to the browser binary automatically on Windows and Linux.

Closes #59.

aksiksi commented on src/cookies.rs:
```diff
@@ -0,0 +1,123 @@
+pub type Cookie<'a> = cookie::Cookie<'a>;
+
+/// JSON representation of a cookie as [defined by WebDriver](https://www.w3.org/TR/webdriver1/#cookies).
+#[derive(Debug, Deserialize, Serialize)]
```
It might make sense to expose this type directly instead of converting to/from a `cookie::Cookie`.  The problem with using `cookie::Cookie` is that some fields will not be forwarded to WebDriver, even when set by the user (e.g., `same_site`).

aksiksi commented on src/cookies.rs:
```diff
@@ -0,0 +1,123 @@
+
+/// JSON representation of a cookie as [defined by WebDriver](https://www.w3.org/TR/webdriver1/#cookies).
+#[derive(Debug, Deserialize, Serialize)]
+struct JsonCookie {
```
It might make sense to expose this type directly instead of converting to/from a cookie::Cookie. The problem with using cookie::Cookie is that some fields will not be forwarded to WebDriver, even when set by the user (e.g., same_site).

jonhoo replied:
I'd actually specifically like to expose a type that isn't tied to the particular WebDriver standard so that the user doesn't have to work too hard to get to something that "just works" with the ecosystem, so let's keep this an implementation detail for the time being.

I also wonder if we should rename this `WebDriverCookie` instead, since that's what it actually represents — the fact that it's encoded as JSON is sort of neither here nor there.

aksiksi replied:
That makes sense!  I've renamed it to `WebDriverCookie`.

jonhoo reviewed (changes requested):
This looks great, thank you! I just left one small note about the naming of the new type. I'd also prefer if the `binary` change could be in its own PR, just to try and keep the record of changes a little neater, but that's not a requirement.

aksiksi commented:
@jonhoo Thank you for reviewing!  I've reverted the `binary` flag change - will open a new PR for that.  The reason I changed it in the first place is because it doesn't work on Windows (my current dev environment).

jonhoo commented on src/cookies.rs:
```diff
@@ -6,12 +6,12 @@ use webdriver::command::WebDriverCommand;
 use crate::error;
 
-/// Type alias for a `cookie::Cookie`
+/// Type alias for a [cookie::Cookie]
```
You can actually write
```
[`cookie::Cookie`]
```
and the intra-doc links feature will still pick it up!

jonhoo commented on src/cookies.rs:
```diff
@@ -0,0 +1,123 @@
+use crate::error;
+
+/// Type alias for a [cookie::Cookie]
+pub type Cookie<'a> = cookie::Cookie<'a>;
```
Out of interest — why did you want to explicitly re-export the `Cookie` type here? Does that make a particular use-case easier?

aksiksi replied:
Ah, I was thinking ahead a bit for the `add_cookie()` API!

My idea was to expose the `cookie::Cookie` that `fantoccini` pulls in so that users can construct one without pulling in the crate themselves.  In addition, this would help reduce the chance of users trying to use a more recent/non-compatible version of the `cookie` crate by mistake.

But I could also be misunderstanding how Cargo handles multiple versions of a dependency!

jonhoo replied:
No, that seems totally reasonable! Good call :+1:

jonhoo commented:
I'll release this once CI kicks back in :)

jonhoo commented:
I don't understand what's going on with the Travis build credits here.. Still keeping an eye on it.

jonhoo commented:
Finally published as 0.17.4 :tada:

aksiksi commented:
Awesome, thanks for following up!
